File: AI-ML/water_jug_problem.c
// #include <stdio.h>
// #include <stdbool.h>

// // Function prototypes
// void waterJugProblem(int x, int y, int z);

// int gcd(int a, int b) {
//     // Find the greatest common divisor
//     if (b == 0) return a;
//     return gcd(b, a % b);
// }

// void waterJugProblem(int x, int y, int z) {
//     if (z > x && z > y) {
//         printf("No solution possible\n");
//         return;
//     }
//     if (z % gcd(x, y) != 0) {
//         printf("No solution possible\n");
//         return;
//     }

//     int a = 0, b = 0; // Initial state
//     while (a != z && b != z) {
//         if (a == 0) {
//             // Fill jug X
//             printf("Fill jug X\n");
//             a = x;
//         } else if (b == y) {
//             // Empty jug Y
//             printf("Empty jug Y\n");
//             b = 0;
//         } else {
//             // Pour water from X to Y
//             int pour = (b + a > y) ? y - b : a;
//             a -= pour;
//             b += pour;
//             printf("Pour %d liters from X to Y\n", pour);
//         }
//         // Print the current state
//         printf("Jug X: %d liters, Jug Y: %d liters\n", a, b);
//     }
//     printf("Target %d liters achieved!\n", z);
// }

// int main() {
//     int x = 4; // Capacity of jug X
//     int y = 3; // Capacity of jug Y
//     int z = 2; // Target amount
//     waterJugProblem(x, y, z);
//     return 0;
// }

File: AI-ML/water_jug_problem.py
from collections import defaultdict

def waterJugSolver(amt1, amt2, jug1, jug2, aim, visited):
    # Print the current state and check the stopping condition
    if (amt1, amt2) in visited:
        return False
    
    print(amt1, amt2)
    if amt1 == aim or amt2 == aim:
        return True
    
    visited.add((amt1, amt2))
    
    # Explore all possible moves recursively in the correct order
    return (waterJugSolver(0, amt2, jug1, jug2, aim, visited) or  # Empty Jug 1
            waterJugSolver(amt1, 0, jug1, jug2, aim, visited) or  # Empty Jug 2
            waterJugSolver(jug1, amt2, jug1, jug2, aim, visited) or  # Fill Jug 1
            waterJugSolver(amt1, jug2, jug1, jug2, aim, visited) or  # Fill Jug 2
            waterJugSolver(amt1 - min(amt1, jug2 - amt2), amt2 + min(amt1, jug2 - amt2), jug1, jug2, aim, visited) or  # Pour Jug 1 -> Jug 2
            waterJugSolver(amt1 + min(amt2, jug1 - amt1), amt2 - min(amt2, jug1 - amt1), jug1, jug2, aim, visited))  # Pour Jug 2 -> Jug 1

# Input values
jug1 = int(input("Enter the capacity of Jug1: "))
jug2 = int(input("Enter the capacity of Jug2: "))
aim = int(input("Enter the amount of water to be measured: "))

# Visited states to avoid redundant checks
visited = set()

# Solve the problem starting from (0, 0)
print("Steps:")
if not waterJugSolver(0, 0, jug1, jug2, aim, visited):
    print("No solution possible.")
